### PR TITLE
Inspector business logic changes to support tracking local widget

### PR DESCRIPTION
**Pull request by jacob314**
class names.

### CODE REVIEW

File: packages/devtools_app/lib/src/app.dart [hunk 1]
```diff
@@ -527,7 +528,10 @@ class SettingsDialog extends StatelessWidget {
 /// Conditional screens can be added to this list, and they will automatically
 /// be shown or hidden based on the [Screen.conditionalLibrary] provided.
 List<DevToolsScreen> get defaultScreens => <DevToolsScreen>[
-      const DevToolsScreen(InspectorScreen(), createController: null),
+      DevToolsScreen<InspectorSettingsController>(
+        const InspectorScreen(),
+        createController: () => InspectorSettingsController(),
```
Comment (jacob314) [inline packages/devtools_app/lib/src/app.dart:533]: not really used until followup animation ui CL.

File: packages/devtools_app/lib/src/inspector/inspector_controller.dart [hunk 2]
```diff
@@ -755,19 +773,15 @@ class InspectorController extends DisposableController
     return null;
   }
 
-  void syncSelectionHelper(
-      bool maybeRerootDetailsTree, RemoteDiagnosticsNode detailsSelection) {
-    if (!detailsSubtree && selectedNode != null) {
```
Comment (jacob314) [inline packages/devtools_app/lib/src/inspector/inspector_controller.dart:760]: this is obsolete with the flutter inspector UI.

File: packages/devtools_app/lib/src/inspector/inspector_service.dart [hunk 3]
```diff
@@ -10,13 +10,19 @@ import 'dart:async';
 import 'dart:convert';
 import 'dart:developer';
 
+import 'package:devtools_app/src/octicons.dart';
+import 'package:flutter/cupertino.dart';
```
Comment (jacob314) [inline packages/devtools_app/lib/src/inspector/inspector_service.dart:14]: This class couldn't really operate without package:flutter so might as well take advantage of package:flutter for handy constants. We now use the unittest runner to run this test on the regular flutter bots.

File: packages/devtools_app/lib/src/inspector/inspector_service.dart [hunk 4]
```diff
@@ -968,11 +1122,73 @@ class ObjectGroup {
...
     };
-    return parseDiagnosticsNodeDaemon(invokeServiceMethodDaemonParams(
+    final json = await invokeServiceMethodDaemonParams(
       'getDetailsSubtree',
       args,
+    );
+    return parseDiagnosticsNodeHelper(json);
+  }
+
+  Future<void> invokeSetFlexProperties(
```
Comment (jacob314) [inline packages/devtools_app/lib/src/inspector/inspector_service.dart:1132]: These methods were all just moved from the separate obsolete extension method file.

File: packages/devtools_app/lib/src/inspector/inspector_service_polyfill.dart [hunk 5]
```diff
@@ -13,7 +13,19 @@ Future<String> loadPolyfillScript() {
   return asset.loadString('assets/scripts/inspector_polyfill_script.dart');
 }
 
-Future<void> invokeInspectorPolyfill(ObjectGroup group) async {
+final _polyfillReadyExpando = Expando<Future>();
```
Comment (jacob314) [inline packages/devtools_app/lib/src/inspector/inspector_service_polyfill.dart:16]: Fix an edge case where we could try to add the polyfill multiple times.

File: packages/devtools_app/test/support/utils.dart [hunk 6]
```diff
@@ -121,3 +124,60 @@ extension RichTextChecking on CommonFinders {
...
+
+/// Workaround to initialize the live widget binding with assets.
+///
+/// The [LiveTestWidgetsFlutterBinding] is useful for unit tests that need to
+/// perform true async operations such as communicating with the VM Service.
+/// Unfortunately the default implementation doesn't work with the patterns we
+/// use to load assets in the devtools application.
+/// TODO(jacobr): consider writing proper integration tests instead rather than
+/// using this code path.
+void initializeLiveTestWidgetsFlutterBindingWithAssets() {
```
Comment (jacob314) [inline packages/devtools_app/test/support/utils.dart:136]: This is the key to running tests we would normally run using a DartVM using flutter test without running into trouble.

File: packages/devtools_app/lib/src/inspector/inspector_service.dart [hunk 7]
```diff
@@ -179,11 +201,115 @@ class InspectorService extends DisposableController
...
+
+      final parts =
+          directory.split('/').where((element) => element.isNotEmpty).toList();
+      final libIndex = parts.lastIndexOf('lib');
+      final path = libIndex > 0 ? parts.sublist(0, libIndex) : parts;
+      // Special case handling of bazel packages.
+      final google3Index = path.lastIndexOf('google3');
+      if (google3Index != -1 && google3Index + 1 < path.length) {
+        var packageParts = path.sublist(google3Index + 1);
+        //
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/inspector/inspector_service.dart:233]: delete empty comment

File: packages/devtools_app/lib/src/inspector/inspector_service.dart [hunk 8]
```diff
@@ -179,11 +201,115 @@ class InspectorService extends DisposableController
...
+        if (packageParts[0] == 'third_party' && packageParts.length >= 3) {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/inspector/inspector_service.dart:234]: add a const for 3 so it is clear what that number is for

File: packages/devtools_app/lib/src/inspector/inspector_service.dart [hunk 9]
```diff
@@ -179,11 +201,115 @@ class InspectorService extends DisposableController
...
+    }
+    return false;
+  }
+
+  @visibleForTesting
+  final Map<String, ClassRef> localClasses = {};
+
+  bool isLocalClass(RemoteDiagnosticsNode node) {
+    if (node.widgetRuntimeType == null) return false;
+    final rawType = node.widgetRuntimeType.split('<').first;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/inspector/inspector_service.dart:290]: add an example of what `node.widgetRuntimeType` should look like in a comment

File: packages/devtools_testing/lib/inspector_service_test.dart [hunk 10]
```diff
@@ -83,12 +82,194 @@ Future<void> runInspectorServiceTests(FlutterTestEnvironment env) async {
...
+            isFalse,
+          );
+          expect(
+            inspectorService.isLocalUri(
+              'package:bar.cores/src/bar.dart',
+            ),
+            isFalse,
+          );
+
+          // Test bazel project like cases.
```
Comment (kenzieschmoll) [inline packages/devtools_testing/lib/inspector_service_test.dart:167]: can this be put in a separate test case? `test('local classes for bazel projects', () {...});`

**Comment by kenzieschmoll** [comment]
done with first round of comments

File: packages/devtools_app/assets/scripts/inspector_polyfill_script.dart [hunk 11]
```diff
@@ -248,6 +252,13 @@ String addServiceExtensions() {
     return Future.value(<String, Object>{'result': succeed});
   }
 
+  Future<Map<String, dynamic>> getPubRootDirectories(
+      Map<String, String> parameters) {
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/devtools_app/assets/scripts/inspector_polyfill_script.dart:256]: done

File: packages/devtools_app/lib/src/inspector/diagnostics_node.dart [hunk 12]
```diff
@@ -78,6 +78,17 @@ class RemoteDiagnosticsNode extends DiagnosticableTree {
...
   Map<String, Object> get size => json['size'];
 
+  bool _isLocalClass;
+
+  bool get isLocalClass {
+    final objectGroup = inspectorService;
+    if (objectGroup is ObjectGroup) {
+      return _isLocalClass ??= objectGroup.inspectorService.isLocalClass(this);
+    } else {
+      return false;
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/devtools_app/lib/src/inspector/diagnostics_node.dart:88]: done. also added a TODO

File: packages/devtools_app/lib/src/inspector/inspector_controller.dart [hunk 13]
```diff
@@ -362,15 +375,23 @@ class InspectorController extends DisposableController
     maybeLoadUI();
   }
 
+  List<String> _rootDirectories;
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/devtools_app/lib/src/inspector/inspector_controller.dart:378]: @pq not sure why I miss this issue when scanning over my own code. Is there a lint we can add to enforce the flutter style for getter, field, setter order?

File: packages/devtools_app/lib/src/inspector/inspector_controller.dart [hunk 14]
```diff
@@ -755,19 +773,15 @@ class InspectorController extends DisposableController
...
-      bool maybeRerootDetailsTree, RemoteDiagnosticsNode detailsSelection) {
-    if (!detailsSubtree && selectedNode != null) {
-      inspectorTree.nodeChanged(selectedNode.parent);
-    }
-    final RemoteDiagnosticsNode diagnostic = selectedDiagnostic;
-    if (diagnostic != null) {
-      if (diagnostic.isCreatedByLocalProject) {
-        _navigateTo(diagnostic);
+  void syncSelectionHelper(bool maybeRerootDetailsTree,
+      RemoteDiagnosticsNode selection, RemoteDiagnosticsNode detailsSelection) {
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/devtools_app/lib/src/inspector/inspector_controller.dart:777]: good idea. done.

File: packages/devtools_app/lib/src/inspector/inspector_service.dart (hunk 7 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/inspector/inspector_service.dart:233]: done

File: packages/devtools_app/lib/src/inspector/inspector_service.dart (hunk 8 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/inspector/inspector_service.dart:234]: done. Added a comment as well.

File: packages/devtools_app/lib/src/inspector/inspector_service.dart (hunk 9 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/inspector/inspector_service.dart:290]: done.

File: packages/devtools_testing/lib/inspector_service_test.dart (hunk 10 above)
Comment (jacob314) [reply packages/devtools_testing/lib/inspector_service_test.dart:167]: done

**Comment by jacob314** [comment]
Done with first round of code review comment fixes.

File: packages/devtools_app/lib/src/inspector/inspector_controller.dart [hunk 15]
```diff
@@ -497,7 +497,10 @@ class InspectorController extends DisposableController
       RemoteDiagnosticsNode detailsSelection, bool setSubtreeRoot) {
     newSelection ??= selectedDiagnostic;
     setSelectedNode(findMatchingInspectorTreeNode(newSelection));
-    syncSelectionHelper(setSubtreeRoot, newSelection, detailsSelection);
+    syncSelectionHelper(
+        maybeRerootDetailsTree: setSubtreeRoot,
+        selection: newSelection,
+        detailsSelection: detailsSelection);
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/inspector/inspector_controller.dart:503]: trailing comma

**Review by kenzieschmoll — APPROVED** [review]
lgtm_orangatang

File: packages/devtools_app/lib/src/inspector/inspector_controller.dart (hunk 15 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/inspector/inspector_controller.dart]: done
